### PR TITLE
Add or Update validation rule into the checkout address shipping and billing form fields

### LINKED ISSUE #8942: Add or Update validation rules into the checkout address form fields.
# New topic request

## Description

The Customize Checkout section has lots of topics to discuss the customization of checkout functionality. PFA.

checkout 

But there is no information about change or update checkout address field validation rule. 

This topic refers to add or update the validation rules into the checkout address form fields.

## Possible solutions

Add instruction to override validation field rules.

### PR DESCRIPTION
**Pull request by rajesh-mage**
## Purpose of this pull request

This pull request (PR) Add or Update validation rule into the checkout address shipping and billing form fields

## Affected DevDocs pages

-  New Topic

Fixes #8942

### CODE REVIEW

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+subgroup:
+title: Add custom validation rules to checkout address form
+subtitle: Customize Checkout
+menu_order: 101
+level3_subgroup: checkout-tutorial
+functional_areas:
+  - Checkout
+---
+
+This topic describes how to add custom address validation in to the Magento checkout.  By default it has own set of validations for the address fields in the checkout step. But sometimes you may want to add some custom validations in to the different address fields. Some examples may be like limiting the character limit in the firstname and lastname of shipping and billing address.
```
Comment (BarnyShergold) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:13]:
```suggestion
This topic describes how to add custom address validation to the Magento checkout.  By default it has own set of validations for the address fields in the checkout step. But sometimes you may want to add some custom validations to the different address fields. One example may be limiting the character limit in the firstname and lastname of the shipping and billing addresses.
```

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+To change existing address validations to the Magento checkout, do the following steps:
```
Comment (BarnyShergold) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:15]:
```suggestion
To change the existing address validations to the Magento checkout, do the following steps:
```

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+1. [Create a new module](#create-module).
+1. [Create a di.xml file](#add-dixml-file).
+1. [Adding validation to Shipping Address](#add-shipping-validation).
```
Comment (BarnyShergold) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:19]:
```suggestion
1. [Adding validation to the Shipping Address](#add-shipping-validation).
```

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+1. [Adding validation to Billing Address](#add-billing-validation).
```
Comment (BarnyShergold) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:20]:
```suggestion
1. [Adding validation to the Billing Address](#add-billing-validation).
```

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+```php?start_inline=1
+<?xml version="1.0"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
+    <type name="Magento\Checkout\Block\Checkout\LayoutProcessor">
+        <plugin name="Vendor_AddressRestriction_Checkout" type="Vendor\AddressRestriction\Block\LayoutProcessor" sortOrder="100"/>
+    </type>
+</config>
+```
+
+## Step 3 Adding validation to Shipping Address {#add-shipping-validation}
```
Comment (BarnyShergold) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:39]:
```suggestion
## Step 3 Adding validation to the Shipping Address {#add-shipping-validation}
```

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
This is a great new addition but there are some omissions and changes in addition to the corrections - 
1. As the 2.4 version of the file is identical to the 2.3 version, please change the 2.4 file to a symlink. You can get the information here : https://github.com/magento/devdocs/wiki/Symlinks

2. There is no menu point addition. There needs to be an entry in the src/_data/toc/how-do-i.yml file

3. Consider adding a link to the bottom of v2.3/howdoi/checkout_overview.md file

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    * @param \Magento\Checkout\Block\Checkout\LayoutProcessor $subject
+    * @param array $jsLayout
+    * @return array
+    */
+    public function afterProcess(
+    \Magento\Checkout\Block\Checkout\LayoutProcessor $subject,
+    array $jsLayout
+    ) {
+        // Firstname
+        $jsLayout['components']['checkout']['children']['steps']['children']['shipping-step']['children']
```
Comment (eduard13) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:61]: In my opinion it's much better to suggest using the [Array Manager](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/framework/array-manager.html), which is much preferred way of accessing a chained array. As result we'll have a much cleaner solution and code.

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md (hunk 5 above)
Comment (rajesh-mage) [reply src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md]: @BarnyShergold Thanks for your review and suggestion. All the suggestions are done please review it.

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md (hunk 6 above)
Comment (rajesh-mage) [reply src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md]: @eduard13 Thanks for your review and suggestion. I have updated the mentioned suggestion. Kindly look into that.

Comment (eduard13) [reply src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md]: @rajesh-mage, looks like you didn't get me right. I suggest using the Array Manager to access the array properties, please check the 1st example in the doc's article.

Please let me know if there are any concerns or questions.

**Comment by dobooth** [comment]
@rajesh-mage Looks like some merge conflicts have appeared. @BarnyShergold is this approved?

**Review by BarnyShergold — APPROVED** [review]

**Comment by BarnyShergold** [comment]
> @rajesh-mage Looks like some merge conflicts have appeared. @BarnyShergold is this approved?

APologies for the delay - this got missed. All good for me but @rajesh-mage, you need to resolve the conflicts

**Comment by keharper** [comment]
I cleaned up the merge conflicts and did some basic cleanup. @rajesh-mage, be sure to pull in those changes.

I'm not sure if you addressed @eduard13's concerns, but I have several more.

* Are steps 3 and 4 really one step? Are you adding code to ``app/code/Vendor/AddressRestriction/Block/LayoutProcessor.php` in each step? If yes, combine the steps. If no, add a sentence under the step 4 heading, stating the file name. Also make sure the code sample is complete.
* Where are the default lengths defined? I see in the `customer_address_entity` table that most of the columns that hold these values have a length of 256. I would like the reader to know how to determine what they're changing the value from.

File: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        return $fields;
+    }
+}
+```
+
+## Step 4: Add validation for the billing address {#add-billing-validation}
+
+```php?start_inline=1
+$fields = $jsLayout['components']['checkout']['children']['steps']['children']['billing-step']['children']
+        ['shippingAddress']['children']['shipping-address-fieldset']['children']['firstname']['validation']['max_text_length'] = 20;
```
Comment (keharper) [inline src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md:103]: Are all these references to the shipping address applicable for validating the billing address? Or is this doing something to ensure the values are the same for both types of addresses?

**Comment by BarnyShergold** [comment]
@rajesh-mage - We've not heard from you on this PR for a week now - can you give us an update in the next few days please?

**Comment by dobooth** [comment]
Haven't heard from @rajesh-mage in over 2 months. Closing as abandoned. Please reopen if you wish to get this merged.

**Pull request closed without merging**